packages/inference/src/HfInference.ts

```diff
@@ -899,4 +993,88 @@ export class HfInference {
 		}
 		return output;
 	}
+
+	/**
+	 * Make request that uses server-sent events and returns response as a stream
+	 */
+	public async streamingRequest<T>(
+		args: Args & { data?: Blob | ArrayBuffer },
+		options?: Options & {
+			binary?: boolean;
+			blob?: boolean;
+			/** For internal HF use, which is why it's not exposed in {@link Options} */
+			includeCredentials?: boolean;
+		}
+	): Promise<ReadableStream<T>> {
+		const mergedOptions = { ...this.defaultOptions, ...options };
+		const { model, ...otherArgs } = args;
+
+		const headers: Record<string, string> = {};
+
+		headers["Accept"] = "text/event-stream";
+
+		if (this.apiKey) {
+			headers["Authorization"] = `Bearer ${this.apiKey}`;
+		}
+
+		if (!options?.binary) {
+			headers["Content-Type"] = "application/json";
+		}
+
+		if (options?.binary) {
+			if (mergedOptions.wait_for_model) {
+				headers["X-Wait-For-Model"] = "true";
+			}
+			if (mergedOptions.use_cache === false) {
+				headers["X-Use-Cache"] = "false";
+			}
+			if (mergedOptions.dont_load_model) {
+				headers["X-Load-Model"] = "0";
+			}
+		}
+
+		const response = await fetch(`${HF_INFERENCE_ENDPOINT_BASE_URL}${model}`, {
+			headers,
+			method: "POST",
+			body: options?.binary
+				? args.data
+				: JSON.stringify({
+						...otherArgs,
+						stream: true,
+						options: mergedOptions,
+				  }),
+			credentials: options?.includeCredentials ? "include" : "same-origin",
+		});
+
+		if (mergedOptions.retry_on_error !== false && response.status === 503 && !mergedOptions.wait_for_model) {
+			return this.request(args, {
+				...mergedOptions,
+				wait_for_model: true,
+			});
+		}
+		if (!response.ok) {
+			throw new Error(`Server response contains error: ${response.status}`);
+		}
```

Comment:
Extracted most of it into options preparation helper. For error/re-try logic, it will be quite different if we switch to `fetch-event-source` someday, so left it in individual methods for now.

Reply:
Thanks!